tests/multi_net/asyncio_tcp_tls_server_client.py

```diff
@@ -0,0 +1,132 @@
+
+try:
+    import os
+    import asyncio
+    import ssl
+except ImportError:
+    print("SKIP")
+    raise SystemExit
+
+PORT = 8000
+PORT_TLS = 8004
+
+# These are test certificates. See tests/README.md for details.
+cert = cafile = "multi_net/rsa_cert.der"
+key = "multi_net/rsa_key.der"
+
+try:
+    os.stat(cafile)
+    os.stat(key)
+except OSError:
+    print("SKIP")
+    raise SystemExit
+
+
+async def handle_tcp_connection(reader, writer):
+    # Test that peername exists (but don't check its value, it changes)
+    writer.get_extra_info("peername")
+
+    data = await reader.read(100)
+    print("echo:", data)
+    writer.write(data)
+    await writer.drain()
+
+    print("close")
+    writer.close()
+    await writer.wait_closed()
+
+    print("done")
+    ev.set()
+
+
+async def handle_tls_connection(reader, writer):
+    # Test that peername exists (but don't check its value, it changes)
+    writer.get_extra_info("peername")
+
+    data = await reader.read(100)
+    print("echo:", data)
+    writer.write(data)
+    await writer.drain()
+
+    print("close")
+    writer.close()
+    await writer.wait_closed()
+
+    print("done")
+    ev1.set()
+
+
+async def tcp_server():
+    global ev
+    ev = asyncio.Event()
+    server = await asyncio.start_server(handle_tcp_connection, "0.0.0.0", PORT)
+    print("tcp server running")
+
+    multitest.next()
+    async with server:
+        await asyncio.wait_for(ev.wait(), 10)
+
+
+async def tcp_client(message):
+    await asyncio.sleep(1)
+    reader, writer = await asyncio.open_connection(IP, PORT)
+    print("write:", message)
+    writer.write(message)
+    await writer.drain()
+    data = await reader.read(100)
+    print("read:", data)
+
+
+async def tls_server():
+    global ev1
+
+    server_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
+    server_ctx.load_cert_chain(cert, key)
+    ev1 = asyncio.Event()
+    server = await asyncio.start_server(handle_tls_connection, "0.0.0.0", PORT_TLS, ssl=server_ctx)
+    print("tls server running")
+    multitest.next()
+    async with server:
+        await asyncio.wait_for(ev1.wait(), 10)
+
+
+async def tls_client(message):
+    client_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+    client_ctx.verify_mode = ssl.CERT_REQUIRED
+    client_ctx.load_verify_locations(cafile=cafile)
+    await asyncio.sleep(1)
+    reader, writer = await asyncio.open_connection(
+        IP, PORT_TLS, ssl=client_ctx, server_hostname="micropython.local"
+    )
+    print("write:", message)
+    writer.write(message)
+    await writer.drain()
+    data = await reader.read(100)
+    print("read:", data)
+
+
+async def tcp_tls_server():
+    print("TCP SERVER")
+    await tcp_server()
+
+    print("TLS SERVER")
+    await tls_server()
+
+
+async def tcp_tls_client():
+    print("TCP CLIENT")
+    await tcp_client(b"client data")
+    multitest.next()
+    print("TLS CLIENT")
+    await tls_client(b"client data")
```

Comment:
This is doing a non-SSL connection followed by an SSL-connection.  But what exactly is that testing?  This feels more like an integration test rather than a unit test.  Did you see issues with non-SSL followed by SSL connections?

Also there's the corresponding `asyncio_tls_tcp_server_client.py` test which flips the order of these connections.  Is it necessary to have both of theses sets of tests?

If it is necessary to have this kind of test I'd suggest just a single test (not two separate files) that does non-SSL then SSL then non-SSL again.

Reply:
>  But what exactly is that testing? This feels more like an integration test rather than a unit test. Did you see issues with non-SSL followed by SSL connections?

That was prior to https://github.com/micropython/micropython/pull/12224 see https://github.com/micropython/micropython/pull/11897#discussion_r1290401298

But I guess now that it's fixed, they are not necessary any more?

Reply:
Oh, I see.

Now that bug is fixed, testing combinations of non-SSL and SSL is no longer needed.

So the tests here can just be for SSL.